### DISCUSSION TITLE
Model dict export with nested dataclasses

### DISCUSSION THREAD

**Discussion by caraballofelix88** [pydantic/pydantic#2752] [Feature Request]
### TLDR
`BaseModel.dict()` should apply recursively to nested dataclasses, for consistency with `BaseModel.json()`

Hi! It's been a pleasure making use of Pydantic so far.  I've got a minor question/request I hope someone could help me out with.

I’ve got a codebase that makes heavy use of built-in data classes, and I’d like to continue using these as building blocks for a new feature being built out w Pydantic models. Builtin dataclasses work just fine if you just drop them in to a Pydantic model, but are missing some functionality that I expected.

If I try to export a Pydantic model to a dict using `model.dict()`, I would expect its member dataclasses would be converted recursively, in the same way that its sub-models would be. 

ex.
```python
from pydantic import BaseModel
from dataclasses import dataclass
from typing import Optional

class NestedModel(BaseModel):
    x: str
    y: int


@dataclass
class NestedThing:
  a: str
  b: int
  c: Optional[str] = None


class Model(BaseModel):
  name: Optional[str]
  model: NestedModel
  thing: NestedThing


m = Model(model=NestedModel(x="blah blah", y=0), thing=NestedThing(a="yadda yadda", b=1))

print(m.dict())
```

What I would expect:
```
{'name': None, 'model': {'x': 'blah blah', 'y': 0}, 'thing': {"a": 'yadda yadda', "b": 1, "c": None}}
```

What I get:
```
{'name': None, 'model': {'x': 'blah blah', 'y': 0}, 'thing': _Pydantic_NestedThing_27955664(a='yadda yadda', b=1, c=None)}
```

Is this expectation sensible? If not, I'd love to hear why. Either way, would appreciate some feedback or suggestions.

**Comment by PrettyWood** [comment]
Hi @caraballofelix88 

I guess both behaviours can be wanted depending on your use case.
We could maybe imagine a `m.dict(all=True)` or any other kwargs to tackle all "model-like" classes like `dataclass`, `TypedDict` or `NamedTuple` ?
In the meantime you can already implement this code for you
```py
from dataclasses import dataclass
from typing import Any, NamedTuple, Optional, TypedDict, no_type_check

from pydantic import BaseModel as PydanticBaseModel


class BaseModel(PydanticBaseModel):
    @classmethod
    @no_type_check
    def _get_value(cls, v, *args, **kwargs) -> Any:
        if hasattr(v, '__pydantic_model__'):
            values = v._asdict() if isinstance(v, tuple) else v.__dict__
            v = v.__pydantic_model__(**values)
        return super()._get_value(v, *args, **kwargs)


class NestedModel(BaseModel):
    x: str
    y: int


@dataclass
class DcThing:
  a: str
  b: int
  c: Optional[str] = None


class NtThing(NamedTuple):
    a: str
    b: int

class TdThing(TypedDict):
    a: str
    b: int


class Model(BaseModel):
  nested: NestedModel
  dc_thing: DcThing
  nt_thing: NtThing
  td_thing: TdThing


m = Model(
    nested=NestedModel(x="blah blah", y=0),
    dc_thing=DcThing(a="yadda yadda", b=1),
    nt_thing=NtThing(a="yadda yadda", b=1),
    td_thing=TdThing(a="yadda yadda", b=1),
)

assert m.dict() == {
    'nested': {'x': 'blah blah', 'y': 0},
    'dc_thing': {'a': 'yadda yadda', 'b': 1, 'c': None},
    'nt_thing': {'a': 'yadda yadda', 'b': 1},
    'td_thing': {'a': 'yadda yadda', 'b': 1},
}
```
Hope it helps!

**Comment by caraballofelix88** [comment]
Thanks so much for the quick and thoughtful reply!

Now that you mention it, I can see either behavior being equally reasonable. Some kwarg on `m.dict()` to enable propagation through model-like properties would be super helpful. 

Should I open up an issue? Given your tidy solution, it seems like extending the `_get_value()` to support this case would be straightforward (and would save me some boilerplate).
